datafusion/optimizer/src/common_subexpr_eliminate.rs

```diff
@@ -454,136 +435,169 @@ impl CommonSubexprEliminate {
-                .map(LogicalPlan::Projection)
-                .map(Transformed::yes)
-        } else {
-            // TODO: How exactly can the name or the schema change in this case?
-            //  In theory `new_aggr_expr` and `new_group_expr` are either the original expressions or they were crafted via `rewrite_expr()`, that keeps the original expression names.
-            //  If this is really needed can we have UT for it?
-            // Alias aggregation expressions if they have changed
-            let new_aggr_expr = new_aggr_expr
-                .into_iter()
-                .zip(saved_names.into_iter())
-                .map(|(new_expr, saved_name)| saved_name.restore(new_expr))
-                .collect::<Result<Vec<Expr>>>()?;
-            // Since group_expr may have changed, schema may also. Use try_new method.
-            let new_agg = if transformed {
-                Aggregate::try_new(new_input, new_group_expr, new_aggr_expr)?
-            } else {
-                Aggregate::try_new_with_schema(
-                    new_input,
-                    new_group_expr,
-                    new_aggr_expr,
-                    orig_schema,
+                    None => Ok((new_aggr_expr, new_group_expr, input, None)),
+                }
+            })?
+            // Recurse into the new input. this is similar to top-down optimizer rule's
+            // logic.
+            .transform_data(|(new_aggr_expr, new_group_expr, new_input, aggr_expr)| {
+                self.rewrite(new_input, config)?.map_data(|new_input| {
+                    Ok((
+                        new_aggr_expr,
+                        new_group_expr,
+                        aggr_expr,
+                        Arc::new(new_input),
+                    ))
+                })
+            })?
+            // Try extracting common aggregate expressions and rebuild the aggregate node.
+            .transform_data(|(new_aggr_expr, new_group_expr, aggr_expr, new_input)| {
+                // Extract common aggregate sub-expressions from the aggregate expressions.
+                self.find_common_exprs(
+                    vec![new_aggr_expr],
+                    config,
+                    ExprMask::NormalAndAggregates,
                 )?
-            };
-            let new_agg = LogicalPlan::Aggregate(new_agg);
-
-            Ok(Transformed::new_transformed(new_agg, transformed))
-        }
+                .map_data(|(mut new_aggr_list, common)| {
+                    let rewritten_aggr_expr = new_aggr_list.pop().unwrap();
+
+                    match common {
+                        // If there are common aggregate sub-expressions, then insert a
+                        // projection above the new rebuilt aggregate node.
+                        Some((common_aggr_exprs, mut aggr_list)) => {
+                            let new_aggr_expr = aggr_list.pop().unwrap();
+
+                            let mut agg_exprs = common_aggr_exprs
```

Comment:
This part is basically the same as it was: https://github.com/apache/datafusion/pull/11683/files#diff-351499880963d6a383c92e156e75019cd9ce33107724a9635853d7d4cd1898d0L522